subprojects/core/src/main/java/org/gradle/api/internal/ExperimentalFeatures.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.internal;
+
+public class ExperimentalFeatures {
+    private static final String ENABLE_EXPERIMENTAL_FEATURES = "org.gradle.internal.experimentalFeatures";
+    boolean enabled;
+
+    public boolean isEnabled() {
+        return enabled
+            || System.getProperty(ENABLE_EXPERIMENTAL_FEATURES) != null;
+    }
+
+    public boolean isEnabled(String featureName) {
```

Comment:
:x: I think we need to change the implementation here, for performance reasons. `System.getProperty` is expensive and not thread safe. It's being called for each graph resolution, and will perform the same lookup again and again. One option would be to disallow changing the system property value _during execution of a build_ and perform a lookup when `ExperimentalFeatures` is created.

Reply:
> One option would be to disallow changing the system property value during execution of a build and perform a lookup when ExperimentalFeatures is created.

I agree that we should do something like this. We do not want to change it during a build anyway.

Reply:
Note that this is an undocumented, internal flag at this stage, so I'm not too worried about exact behaviour. But performance is a concern.

_If_ we want to keep the fine-grained control of features via system properties, I guess I can list all properties on startup and build a map of any that start with `org.gradle.internal.experimentalFeatures`. Otherwise, we could remove the fine-grained control and just check for a single property. WDYT?

Reply:
Yes, my point is exactly that: using `System.getProperty` is too expensive for this. I was also thinking we could have an enum inside `ExperimentalFeatures`. Something like:

```
class ExperimentalFeatures {
     EnumMap<ExperimentalFeature, Boolean> features

     ExperimentalFeatures {
          features = new ...
          for (ExperimentalFeature feat: ExperimentalFeature.values()) {
              features.put(feat, feat.isEnabled())
          }
     }
 
    boolean isEnabled(ExperimentalFeature feature) { return features.get(feature); }

     enum ExperimentalFeature {
         optionalDependencies,
         gradleMetadata

         boolean isEnabled() { System.getProperty('org.gradle.internal.experimental.'+this.name() }
     }
}
```

Reply:
I'd rather not have an enum of experimental features, if we can avoid it. The concept of named features was kind of a 'just-in-case' thing that we might find useful if someone reported an issue that we were confused by. But maybe it's actually a good idea? (It's just that we agreed _not_ to give the user too many options, since it's generally confusing).

 One option is to just remove the ability to control separate features until we actually have a use case. Then, all we have to do is check for `org.gradle.internal.experimentalFeatures` on startup.

That's what my WDYT? was asking. Do we even need fine-grained control?

Reply:
Let's go without individual control, until we have a good use case.

Reply:
Done: https://github.com/gradle/gradle/pull/3479/commits/e06cce397db8f8d8de7c891aa188cbd5fe1eb3ea and https://github.com/gradle/gradle/pull/3479/commits/3c2b71f56b1c7e4df180ceaa3c4160e9dc275abe